### ISSUE TITLE
sparse gradient update

### ISSUE THREAD

**Issue by chunyang-wen** [ray-project/ray#3539]
### System information

Oops ！

### Describe the problem

In ray sgd, `ParameterServer` is responsible for gradients accumulating and delivering. So before creating a `ParameterServer`, we need to know each shape of the gradients (returned by `shard_shapes`). Each `SGDWorker` will hold the same copy of weights, but they will share the gradients. But when it comes to sparse variables,

+ Creating a large dense matrix for sparsely generated gradients is memory consuming
+ Dense updating sparse variable has performance issues.
+ Under certain situations, parameters cannot be fitted into a single machine.

Also, `shard_shape` may be not working for all tensorflow types.

If not stated clearly, @tengteng please help clarify it.

@ericl @alexbao @tengteng

**Comment by tengteng** [comment]
Thanks @chunyang-wen 
I tried to do tf.convert_to_tensor() for sparse IndexedSlices -> dense Tensor. It works but steps very slow since we have million-level sparse features.
We want to have sparse_add()/sparse_get() (or something equivalent) for communicating only the updated features rather than all features in every ps iteration.

**Comment by ericl** [comment]
Hm, it's probably not too hard to change the parameter server actor to allow sparse gradients. Will it work with the sum_gradients_all_reduce() code for the same-machine allreduce though?

**Comment by robertnishihara** [comment]
So one thing we'll need is an efficient way to serialize sparse tensors. E.g., a custom serializer/deserializer like the folowing pseudocode.

```python
def serialize(sparse_tensor):
    return sparse_tensor.values(), sparse_tensor.indices(), sparse_tensor.dtype()

def deserialize(serialized_sparse_tensor):
    return tf.SparseTensor(serialized_sparse_tensor[0], serialized_sparse_tensor[1], serialized_sparse_tensor[2])
```

Do you know if there's a good way to do something like that?

We probably won't be able to use the custom plasma op.

**Comment by ericl** [comment]
It looks like a sparse tensor is just three dense tensors, so that part doesn't seem hard. It should work with the plasma op just fine.

**Comment by piaoyats** [comment]
Let me describe the problem
1. Distributed Session
    Suppose the Wind_and_Deep model, the variable has a large shape the can not place on one worker, in Tensorflow we use the ps and worker, the large variable will shard and place on different ps. Each worker hold the some model, read different train set shard for training,  ps update the weight by gradients in synchronization or asynchronous mode。 How can we setup this distributed model here？ Seem DistributedSGD is in synchronization mode and useful for CNN which can hold the total weight in one host,it use tensorflow local session for training, sgd_step to _average_gradients and update each worker's  model.

2. Support KVVariable for embedding
   Now Tensorflow‘s variable is a dense tensor, we should define the shape before we build model。Actually  the input features for embedding layer (tf.embedding_lookup_sparse) are always discrete characteristics，such as user_id, item_id, in this case we need specific the weight's shape for embedding [vocab_size, embedding_dim], but the vocab_size will change for new unseen user_id coming.   Tensorflow use category column to support these features, it use hash bucket to map a id to a bucket id, this may has Hash conflict because different id may map to a same bucket id。 We implemented a KVVariable in our internal tensorflow version to solve this problem, 
we only need to specific the embedding dim size, the underly implement we use hash_map<T, Tensor> to store the weight, so new unseen feature can be insert into the table with key is the new feature and value is a vector from initialization. All optimizer can update this table with training gradients. 


We want these demands can implement in ray sgd, do we have any idea?

**Comment by tengteng** [comment]
@robertnishihara  @ericl 
The comment above is from one of our teammate who worked on our internal Tensorflow and another Parameter Server system. His comment basically introduced the background(e.g, KVVariable) and the problem we are facing when we are trying to migrate the system to DistributedSGD(Large variable/Hash Sparse features). Please take a look. All comments are welcome. Thanks!

**Comment by ericl** [comment]
1. If I understand correctly, the main issue is that workers cannot store the weights. Since the current Ray SGD does not have a true PS but instead has a "gradient server", this is a problem due to the memory consumption, and also that the gradient aggregation strategy only minimally reduces the size of sparse updates (so you end up with all-to-all gradient broadcast).
I'm not sure if this is easy to solve without moving to more of a true PS-like strategy with the weights off the workers. For your sparse weights, is it possible to the needed ones for the forward pass in a few RPCs? Maybe we can use the PS actors as true PS shards and hold the weights there.

2. It sounds like you need a global hash map that is periodically updated. Could this be implemented as another Ray actor?

**Comment by tengteng** [comment]
@ericl for the 2nd question it'll be a Key-value (feature-id/hash -> feature value) store that is periodically updated (consider online learning scenario when new feature comes in or absolute features get kicked out...) But I'm not sure it should be an actor or should be transparent to SGD (as all developed inside TF). pros & cons?

For the 1st question if we move & shard weights into SGD PS actor:
1. What's the benefit compare with using ray to kick off a distributed TF - using built-in tf ps? To me the differences are: A. gRPC vs plasma in **communication**; B. ray actor **fail over** feature.
2. If we put weight into PS actors. The next question will be how do we shard weights on ps_list, and pass this information to the corresponding workers (or all workers). I feel this process is just like re-implement the same logic that already exists in distributed TF..

**Comment by piaoyats** [comment]
@ericl If we consider integrate distributed TF in Ray，We can make full use of the functions of TF, and the solution will be simple. 

I have an idea, we can also integrate pytorch or others in the same way. To making full use of the computing power of these deep learning frameworks, and combined with ray's flexibility, FO, rllib and raytune, the ML ecosystem of ray can be constructed.
